infrasec/good-infra.md

```diff
@@ -0,0 +1,170 @@
+
+## Minimalism
+
+Minimalist infrastructure is the practice of building only what you need with the fewest number of resources. This is a philosophy, not a religion: don't build bad infrastructure to achieve a minimalist design. A minimalist design should lead to good infrastructure by reducing the complexity of the design and the number of resources being managed. A simpler system with less resources has a smaller attack surface and is easier for people to learn and debug.
+
+### Avoid state
+
+When designing your system, avoid storing additional state. Often the data you want to store is already available in the system. Using the system as the source of truth can avoid the difficult business of ensuring data consistency.
+
+| Practical Example
+| ---
+| Let's say you want to be able to rollback a Fargate deploy if the new task definition results in a service that won't become healthy. One option would be store the working task definition in something like DynamoDB (or git or any number of bad choices). However, your ECS service already has this information: the previous, healthy service is still running. Instead of managing a DynamoDB resource as well as writing code for maintaining that state record, query the ECS service itself to get what you want. This avoids managing the DynamoDB resource as well as the unfortunate scenario where the data stored is wrong.
+
+### Embrace immutability
+
+Related to avoiding state, making resources immutable reduces the scope of what can change. The less that can change, the easier a system is to understand and debug and the more likely that each instance will behave as expected. Reduce systems and architectures  to testable units.
+
+As an illustration, consider the complexity difference between a read-only, distroless, container image with a single application binary and a long-lived VM that's updated with the latest application code and it's dependency tree. How confident are you in each system? How confident are you that prod and dev are similar? What does a delivery pipeline look like for each?
+
+### Use managed services
+
+Manage only what you must. There are lot of services out there that solve infrastructure problems for us. For example, AWS Fargate makes it so you don't have to manage EC2 instances running your containers. This is good. Make managing infrastructure someone else's problem (you've got plenty). Build and maintain the least amount possible.
+
+Examples:
+
+* Parameter Store instead of Hashicorp Vault
+* Fargate instead of EC2
+* SaaS versus on-premise
+
+### Codify entire stacks
+
+Codify entire stacks into a single callable unit (calling into other units). Keep the number of parameterizations (inputs) small so the infrastructure doesn't vary wildly and the unit's actions remain clear. This makes it easy to spin up the same infrastructure in multiple environments and reduces the proliferation of code unique to each environment.
```

Comment:
I wonder if it's worth noting that it's often good to provide the *ability* to input tweaked parameters, but have sensible defaults set. That's one thing I ran into with MilMove, when we had to move everything into GovCloud, is that there were a lot of things that had been hardcoded that would have been a lot easier if there was a way to just tweak the module from the calling point. Obviously there are tradeoffs here, but..

Reply:
Also, there is a discussion to be had here about how large a module gets before it becomes unwieldy. An eye towards being able to modify the code in the future is something that needs to be taken into account.

Reply:
I wonder if this is the place to say something about not encoding the notion of environments or other artificial constructs in the code...

Reply:
Generally agree with what Chas is saying. I think it's a good spot to put some best practice notes and philosophy behind them here. Specifically because we know they come up.

Reply:
I'll take another stab at this. The idea here is to have a lightweight wrapper for a whole stack. I purposely don't use the word "module" because maybe this is a script or tool. The point is to provide an easy way to deploy the same stack to any environment with minimal difference. I've been wondering if this is our (InfraSec) version of a deployable "artifact". That's the kind of thinking I'm doing around it.

Reply:
Alright. I added a little bit about treating these as Twelve-Factor apps. I'm happy to have a bigger discussion and revise accordingly! I'd like to hear what happened on MilMove, that might help me understand what else needs to be said.